### DISCUSSION TITLE
Metrics? eg CPU & memory usage per app

### DISCUSSION THREAD

**Discussion by mariusa** [piku/piku#287] [Announcements]
Hi, what would you recommend to get something similar to Heroku metrics per app?
eg history of CPU & memory usage per app

Also, would it be possible to have usage quotas per app? I wouldn't want to have an app stuck at 100% CPU completely slow down everything else on the system. Or maybe it's not needed, the OS handles this?

PS: Would you enable "Discussions" in github for these kinds of questions?

Thanks!

**Comment by rcarmo** [comment]
Hi! Yeah, enabling Discussions seems like a good idea. At least it's more searchable than Discord :)

There is no built-in way to have CPU and history per app - yet. both @chr15m and myself have taken a few stabs at this in the past (I have a little monitor app that looks at process names and CPU counters, but never cleaned it up for publishing), but most of the time it just doesn't add a lot of value when you're running on small machines.

As to usage quotas, I have been looking at using `cgroups` for that, but it requires a lot more moving parts (I have an experimental setup [here](https://github.com/piku/cloud-init/blob/master/azure/cloud-init.yml#L136) that I have been playing with). But, again, most of the time it just isn't necessary unless you have something that does a lot of batch processing and hogs the machine.

**Comment by mariusa** [comment]
Thank you! I'll leave the request open for metrics, especially since there were some attempts.

**Comment by chr15m** [comment]
@mariusa I built this little hack which just measures CPU every minute and if it is higher than threshold takes a lot of all running processes. It's generally enough to determine which app is eating the CPU: https://github.com/piku/piku-monitoring

**Comment by mariusa** [comment]
Nice, I'll try it out after managing to setup piku. Thanks!

FYI see also https://github.com/Unitech/pm2
but I see that has 700 issues, including https://github.com/Unitech/pm2/issues/5559
Seems like they want to push their own proprietary & centralized system through this tool.

**Comment by rcarmo** [comment]
Also moving this to discussions for archival purposes (follow ups can go there)
